language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/FromJsonGenerator.scala

```diff
@@ -0,0 +1,260 @@
+      fields.zipWithIndex.foreach { case (f, i) =>
+        // We generate `JsonLfReader.Field` as a literal as $T seems to always use fully qualified name.
+        block.addStatement(
+          "case $S: return JsonLfReader.Decoders.Field.at($L, $L)",
+          f.javaName,
+          i,
+          jsonDecoderForType(f.damlType),
+        )
+      }
+      block
+        .addStatement("default: return null")
+        .endControlFlow() // end switch
+        .endControlFlow() // end lambda
+        .build()
+    }
+
+    val constr = {
+      val args =
+        (0 until fields.size).map(CodeBlock.of("$T.cast(args[$L])", decodeClass, _))
+      CodeBlock.of("(Object[] args) -> new $T($L)", typeName, CodeBlock.join(args.asJava, ", "))
+    }
+
+    MethodSpec
+      .methodBuilder(methodName)
+      .addModifiers(modifiers: _*)
+      .addTypeVariables(typeParams.map(TypeVariableName.get).asJava)
+      .addParameters(jsonDecoderParamsForTypeParams(typeParams))
+      .returns(decoderTypeName(typeName))
+      .addStatement(
+        "return $T.record($L, $L, $L)",
+        decodeClass,
+        fieldNames,
+        fieldsByName.toString(),
+        constr,
+      )
+      .build()
+  }
+
+  private def fromJsonString(
+      className: ClassName,
+      typeParams: IndexedSeq[String],
+  ): MethodSpec =
+    MethodSpec
+      .methodBuilder("fromJson")
+      .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+      .addTypeVariables(typeParams.map(TypeVariableName.get).asJava)
+      .addParameter(classOf[String], "json")
+      .addParameters(jsonDecoderParamsForTypeParams(typeParams))
+      .returns(className.parameterized(typeParams))
+      .addException(classOf[java.io.IOException])
+      .addStatement(
+        "return jsonDecoder($L).decode(new $T(json))",
+        CodeBlock.join(typeParams.map(t => CodeBlock.of(decodeTypeParamName(t))).asJava, ", "),
+        classOf[JsonLfReader],
+      )
+      .build()
+
+  def forVariant(
+      className: ClassName,
+      typeParams: IndexedSeq[String],
+      fields: Fields,
+  ): Seq[MethodSpec] = {
+    val typeName = className.parameterized(typeParams)
+
+    val tagNames = CodeBlock.of(
+      "$T.asList($L)",
+      classOf[java.util.Arrays],
+      CodeBlock.join(fields.map(f => CodeBlock.of("$S", f.javaName)).asJava, ", "),
+    )
+    val variantsByTag = {
+      val block = CodeBlock
+        .builder()
+        .beginControlFlow("name ->")
+        .beginControlFlow("switch (name)")
+      fields.foreach { f =>
+        block.addStatement(
+          "case $S: return $L($L)",
+          f.damlName,
+          decoderForTagName(f.damlName),
+          CodeBlock.join(typeParams.map(t => CodeBlock.of(decodeTypeParamName(t))).asJava, ", "),
+        )
+      }
+      block
+        .addStatement("default: return null")
+        .endControlFlow() // end switch
+        .endControlFlow() // end lambda
+        .build()
+    }
+
+    val jsonDecoder = MethodSpec
```

Comment:
if I understand it correctly what you do atm is to generate a `jsonDecoder` and `fromJsonString` method for each type. This definitely is very useful but I think in your current setup it's pretty much impossible to write generic code. We've found that to be extremely useful both in CN and now in the app template.

For reference, here is the code we use to decode JSON blobs to Java codegen types in the app template https://github.com/DACH-NY/daml-app-template/blob/8262896b28c5c6c652e39a6165b608948abcaa51/app/backend/src/main/java/com/daml/app/common/TemplateContract.java#L100,L106.

You can see that at a minimum we need the option to decode from a contract companion (and similar for interfaces). So I suggest we add methods to that for decoding (I'm fine with doing that in a separate PR).

The other point which is a bit more difficult to resolve is that in our case we don't actually have a string. We have what OpenAPI generates for an [object](https://github.com/DACH-NY/daml-app-template/blob/8262896b28c5c6c652e39a6165b608948abcaa51/app/backend/src/main/openapi/api.yaml#L81). This ends up being mapped to a plain Java object. Now as you can see from the code we just convert that to a string again and then decode that but that's obviously not great from a performance perspective. I think the nice option in this case would be jackson’s [convertValue](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/ObjectMapper.html#convertValue(java.lang.Object,%20java.lang.Class). Have you considered supporting that? Afaict that would require providing a jackson deserializer and a module that registers all deserializers. You could probably initially implement this by converting to a string and using your existing deserializers but it allows clients to write the "right" code that we could then later make efficient.

Lastly as you can see from the code above we also need JSON serialization not just deserialization. Definitely out of scope for this PR, just want to put that on your radar.

Reply:
Hmmm, my understanding was that the primary initial usage of the json<->codegen would be in a setup where the data is read from PQS. There, your payload is in json-formatted text. This usage seems to read the data over grpc, which seems like a different case, and is already served by the existing `valueDecoder`s. 
Regarding generic code, perhaps I'm missing something but it looks to me like if your input were text, you could mostly serve the purpose you've shown by passing in a `JsonLfDecoder<Data>` instead of the `companion`, `mapper` and `contractMapper` args. We might need a little more wiring around the contract id, but that can be added.
Decoding directly from string avoids all the intermediate decoding and encoding steps.

Yes, I was expecting that encoding might be needed soon enough. It's strictly simpler than decoding so shouldn't be too much trouble once it's started.

Reply:
So, based on our discussion it sounds like the the needs are:
1. put an additional `fromJson` method on the companion for consistency with the usage of the existing `ValueDecoder`
2. JsonLfEncoder functionality
3. The capacity to decode from an in-memory JSON object, for frameworks where the original JSON has already been decoded and the object embedded by the time you get access to it.

Perhaps I'll do the first one in this PR and create tickets for the other two?

Reply:
sounds good!